File: sandbox/app/persistence/ownership.py
"""Ownership filter helpers for multi-tenant repository SQL."""

from __future__ import annotations

from sandbox.app.domain.types import OwnerScope
from sandbox.app.persistence.errors import OwnershipError


def require_owner_scope(
    scope: OwnerScope | dict[str, str] | None,
    *,
    resource: str | None = None,
) -> OwnerScope:
    """Require non-empty org_id and user_id (plan multi-tenant ownership)."""
    if scope is None:
        raise OwnershipError(
            "Owner scope requires non-empty org_id and user_id",
            resource=resource,
        )
    if isinstance(scope, OwnerScope):
        org_id = str(scope.org_id).strip()
        user_id = str(scope.user_id).strip()
    else:
        org_id = str(scope.get("org_id") or scope.get("orgId") or "").strip()
        user_id = str(scope.get("user_id") or scope.get("userId") or "").strip()
    if not org_id or not user_id:
        raise OwnershipError(
            "Owner scope requires non-empty org_id and user_id",
            resource=resource,
        )
    return OwnerScope(org_id=org_id, user_id=user_id)


def apply_owner_scope_sql(
    where_sql: str,
    scope: OwnerScope,
    *,
    org_column: str = "org_id",
    user_column: str = "user_id",
    table_alias: str | None = None,
) -> tuple[str, tuple[str, str]]:
    """Append org_id + user_id equality predicates.

    Returns ``(sql_fragment, (org_id, user_id))`` for parameterized binding.
    """
    prefix = f"{table_alias}." if table_alias else ""
    owner_clause = (
        f"{prefix}{org_column} = %s AND {prefix}{user_column} = %s"
    )
    base = (where_sql or "").strip()
    if not base:
        return owner_clause, (scope.org_id, scope.user_id)
    if base.upper().startswith("WHERE"):
        return f"{base} AND {owner_clause}", (scope.org_id, scope.user_id)
    return f"{base} AND {owner_clause}", (scope.org_id, scope.user_id)

File: agent/src/infrastructure/mysql/ownership.ts
/**
 * Ownership filter helpers for multi-tenant repository queries.
 */

export type OwnerScope = {
  orgId: string;
  userId: string;
};

/**
 * @param scope
 * @returns {OwnerScope}
 */
export function requireOwnerScope(scope: Partial<OwnerScope> | null | undefined) {
  const orgId = scope?.orgId != null ? String(scope.orgId).trim() : '';
  const userId = scope?.userId != null ? String(scope.userId).trim() : '';
  if (!orgId || !userId) {
    throw new Error(
      'Owner scope requires non-empty orgId and userId (plan §31.11)',
    );
  }
  return { orgId, userId };
}

/**
 * Apply org_id + user_id equality filters to a knex query builder.
 * @param query
 * @param scope
 * @param [cols]
 */
export function applyOwnerScope(query: import('knex').Knex.QueryBuilder, scope: OwnerScope, cols: { orgColumn?: string, userColumn?: string } = {}) {
  const orgColumn = cols.orgColumn || 'org_id';
  const userColumn = cols.userColumn || 'user_id';
  return query.where(orgColumn, scope.orgId).andWhere(userColumn, scope.userId);
}
